src/Microsoft.AspNetCore.SignalR.Client.TS/Formatters.ts

```
@@ -0,0 +1,146 @@
+
+        // parseInt is too leniant, we need a strict check to see if the string is an int
+
+        if (!LengthRegex.test(lenStr)) {
+            return [`Invalid length: '${lenStr}'`, position, null];
+        }
+        let length = Number.parseInt(lenStr);
+
+        // Required space is: 3 (type flag, ":", ";") + length (payload len)
+        if (!hasSpace(input, offset, 3 + length)) {
+            return ["Message is incomplete", position, null];
+        }
+
+        // Read the type
+        var [typeStr, offset] = splitAt(input, ":", offset);
+
+        // Parse the type
+        var type = knownTypes[typeStr];
+        if (type === undefined) {
+            return ["Unknown type value: '" + typeStr + "'", position, null];
+        }
+
+        // Read the payload
+        var payload = input.substr(offset, length);
+        offset += length;
+
+        // Verify the final trailing character
+        if (input[offset] != ';') {
+            return ["Message missing trailer character", position, null]
+        }
+        offset += 1;
+
+        if (type == MessageType.Binary) {
+            // We need to decode and put in an ArrayBuffer. Throw for now
+            // This will require our own Base64-decoder because the browser
+            // built-in one only decodes to strings and throws if invalid UTF-8
+            // characters are found.
+            throw "TODO: Support for binary messages";
+        }
+
+        return [null, offset, new Message(type, payload)];
+    }
+
+    export function parseMessages(input: string): [string, Message[]] {
+        if (input.length == 0) {
+            return [null, []];
+        }
+
+        let messages = [];
+        var offset = 0;
+        while (offset < input.length) {
+            var [error, offset, message] = parseMessage(input, offset);
+            if (error) {
+                // Failed to parse message. Report the error and the messages we parsed so far
```

In that case, we can't really use exceptions for error reporting then because they interrupt the control flow. When I had a version that returned a tuple, we could return the errors as well as the individual messages.

Or we could refactor around a different kind of API that parses messages incrementally, but then we have a lot more state to track internally about where we are in the input stream.